android/modules/ui/src/java/ti/modules/titanium/ui/widget/picker/TiUIDatePicker.java

```diff
@@ -137,58 +214,70 @@ public void onDateChanged(DatePicker picker, int year, int monthOfYear, int dayO
 			targetCalendar.setTime(this.maxDate);
-			setValue(this.maxDate.getTime(), true);
+			setValue(this.maxDate, true);
 		}
 
-		Date newTime = targetCalendar.getTime();
-		Object oTime = proxy.getProperty(TiC.PROPERTY_VALUE);
-		Date oldTime = null;
-
-		if (oTime instanceof Date) {
-			oldTime = (Date) oTime;
+		// Do not continue if proxy was released.
+		if (this.proxy == null) {
+			return;
 		}
 
-		// Due to a native Android bug in 4.x, this callback is called twice, so here
-		// we check if the dates are identical, we don't fire "change" event or reset value.
-		if (oldTime != null && oldTime.equals(newTime)) {
+		// Do not continue if date hasn't changed. (We should only fire event if changed.)
+		Date newTime = targetCalendar.getTime();
+		Date oldTime = TiConvert.toDate(this.proxy.getProperty(TiC.PROPERTY_VALUE));
+		if ((oldTime != null) && oldTime.equals(newTime)) {
 			return;
 		}
 
-		if (!suppressChangeEvent) {
+		// Update "value" property with selected date.
+		this.proxy.setProperty(TiC.PROPERTY_VALUE, newTime);
+
+		// Fire a "change" event.
+		if (!this.suppressChangeEvent) {
 			KrollDict data = new KrollDict();
 			data.put(TiC.PROPERTY_VALUE, newTime);
 			fireEvent(TiC.EVENT_CHANGE, data);
 		}
-
-		proxy.setProperty(TiC.PROPERTY_VALUE, newTime);
 	}
 
-	public void setValue(long value)
+	public void setValue(Date value)
 	{
 		setValue(value, false);
 	}
 
-	public void setValue(long value, boolean suppressEvent)
+	public void setValue(Date value, boolean suppressEvent)
 	{
-		DatePicker picker = (DatePicker) getNativeView();
-		Calendar calendar = Calendar.getInstance();
-		calendar.setTimeInMillis(value);
-		suppressChangeEvent = suppressEvent;
-		picker.updateDate(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH),
-						  calendar.get(Calendar.DAY_OF_MONTH));
-		suppressChangeEvent = false;
-	}
+		// Update DatePicker view if used.
+		DatePicker datePicker = getDatePicker();
+		if (datePicker != null) {
+			Calendar calendar = Calendar.getInstance();
+			calendar.setTimeInMillis(value.getTime());
+			this.suppressChangeEvent = suppressEvent;
+			datePicker.updateDate(
+				calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH));
+			this.suppressChangeEvent = false;
+			return;
+		}
 
-	public void setCalendarView(boolean value)
-	{
-		DatePicker picker = (DatePicker) getNativeView();
-		picker.setCalendarViewShown(value);
+		// We're likely using a text field instead. Format date to localized string.
+		EditText editText = null;
+		View view = getNativeView();
+		if (view instanceof TextInputLayout) {
+			editText = ((TextInputLayout) view).getEditText();
+		} else if (view instanceof EditText) {
+			editText = (EditText) view;
+		}
+		if (editText != null) {
+			DateFormat dateFormat = DateFormat.getDateInstance(DateFormat.MEDIUM);
+			editText.setText(dateFormat.format(value));
+			editText.requestLayout();
```

Comment:
Not sure if this is necessary, I believe `setText` also decides if a relayout is needed

Reply:
It's definitely necessary if the width is set to `Ti.UI.SIZE` and the string gets longer. I noticed this when changing from a single digit hour (say `1:00 AM`) to a digit hour (say `12:00 PM`). A character will get clipped unless I call `requestLayout()`.